src/main/java/io/strimzi/kafka/bridge/mqtt/core/MqttServerHandler.java

```diff
@@ -91,15 +115,82 @@ private void handleConnectMessage(ChannelHandlerContext ctx) {
         ctx.writeAndFlush(connAckMessage);
     }
 
+    /**
+     * Send a MQTT PUBACK message to the client.
+     *
+     * @param ctx ChannelHandlerContext instance
+     * @param packetId packet identifier
+     */
+    private void sendPubAckMessage(ChannelHandlerContext ctx, int packetId) {
+        MqttMessage pubAckMessage = MqttMessageBuilders.pubAck()
+                .packetId(packetId)
+                .build();
+
+        ctx.writeAndFlush(pubAckMessage);
+    }
+
     /**
      * Handle the case when a client sent a MQTT PUBLISH message type.
      *
      * @param ctx            ChannelHandlerContext instance
      * @param publishMessage represents a MqttPublishMessage
      */
     private void handlePublishMessage(ChannelHandlerContext ctx, MqttPublishMessage publishMessage) {
-        logger.info("MQTT Topic: {}", publishMessage.variableHeader().topicName());
-        logger.info("Kafka Mapped Topic: {}", mqttKafkaMapper.map(publishMessage.variableHeader().topicName()));
-        logger.info("Message: {}", publishMessage.payload().toString(Charset.defaultCharset()));
+        // get QoS level from the MqttPublishMessage
+        MqttQoS qos = MqttQoS.valueOf(publishMessage.fixedHeader().qosLevel().value());
+
+        // get the MQTT topic from the MqttPublishMessage
+        String mqttTopic = publishMessage.variableHeader().topicName();
+
+        // perform topic mapping
+        String kafkaMappedTopic = mqttKafkaMapper.map(mqttTopic);
+
+        //log the topic mapping
+        logger.info("MQTT topic {} mapped to Kafka Topic {}", mqttTopic, kafkaMappedTopic);
+
+        // build the Kafka record
+        ProducerRecord<String, byte[]> record = new ProducerRecord<>(kafkaMappedTopic,
+                publishMessage.payload().array());
```

Comment:
It seems not working, I got the following exception ...

```shell
[2023-07-09 15:32:09,537] ERROR <erverHandler:84> [oopGroup-3-1] Error reading Channel: 
java.lang.UnsupportedOperationException: direct buffer
	at io.netty.buffer.PooledUnsafeDirectByteBuf.array(PooledUnsafeDirectByteBuf.java:226) ~[netty-buffer-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.buffer.AbstractPooledDerivedByteBuf.array(AbstractPooledDerivedByteBuf.java:118) ~[netty-buffer-4.1.94.Final.jar:4.1.94.Final]
	at io.strimzi.kafka.bridge.mqtt.core.MqttServerHandler.handlePublishMessage(MqttServerHandler.java:153) ~[classes/:?]
	at io.strimzi.kafka.bridge.mqtt.core.MqttServerHandler.channelRead0(MqttServerHandler.java:80) ~[classes/:?]
	at io.strimzi.kafka.bridge.mqtt.core.MqttServerHandler.channelRead0(MqttServerHandler.java:39) ~[classes/:?]
```

Having a very quick look, it seems something like this is the way to go:

```java
byte[] data = new byte[publishMessage.payload().readableBytes()];
publishMessage.payload().readBytes(data);
// build the Kafka record
ProducerRecord<String, byte[]> record = new ProducerRecord<>(kafkaMappedTopic, data);
```

It's just worth checking if there is any better way.

Reply:
@ppatierno what about using the previous approach?
It was actually doing the same as the code above, plus handling exceptions. I would go with a custom Serializer as it gives the possibility of having single responsibility, thus more readable and maintainable code. I mean, let us not add s many codes not related to the main task of handling an incoming message.

Reply:
No we should stick with the default byte serializer. There is really no reason to have a new one doing exactly the same as the Kafka one just because of the Netty ByteBuf. Let's just get the bytes as Netty allows us.